plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -823,6 +860,88 @@ private void finishOptimize(ConnectorSession session, IcebergTableExecuteHandle
+        int tableVersion = ((BaseTable) table).operations().current().formatVersion();
+        if (tableVersion > EXPIRE_SNAPSHOTS_MAX_SUPPORTED_TABLE_VERSION) {
+            throw new TrinoException(NOT_SUPPORTED, format("EXPIRE_SNAPSHOTS is not supported for iceberg table version > %d. " +
+                            "Table %s version is %s.",
+                    EXPIRE_SNAPSHOTS_MAX_SUPPORTED_TABLE_VERSION,
+                    executeHandle.getSchemaTableName(),
+                    tableVersion));
+        }
+        Map<String, String> properties = table.properties();
+        if (properties.containsKey(WRITE_LOCATION_PROVIDER_IMPL)) {
+            throw new TrinoException(NOT_SUPPORTED, "Table " + executeHandle.getSchemaTableName() + " specifies " + properties.get(WRITE_LOCATION_PROVIDER_IMPL) +
+                    " as a location provider. Writing to Iceberg tables with custom location provider is not supported.");
+        }
+        Duration retention = requireNonNull(expireSnapshotsHandle.getRetentionThreshold(), "retention is null");
+        Duration minRetention = getExpireSnapshotMinRetention(session);
+        checkProcedureArgument(retention.compareTo(minRetention) >= 0,
+                "Retention specified (%s) is shorter than the minimum retention configured in the system (%s). " +
+                        "Minimum retention can be changed with %s configuration property or iceberg.%s session property",
+                retention,
+                minRetention,
+                IcebergConfig.EXPIRE_SNAPSHOTS_MIN_RETENTION,
+                IcebergSessionProperties.EXPIRE_SNAPSHOTS_MIN_RETENTION);
+
+        long expireTimestampMillis = session.getStart().toEpochMilli() - retention.toMillis();
+        expireSnapshots(table, expireTimestampMillis, session, executeHandle.getSchemaTableName());
+    }
+
+    private void expireSnapshots(Table table, long expireTimestamp, ConnectorSession session, SchemaTableName schemaTableName)
+    {
+        Set<String> originalFiles = buildSetOfValidFiles(table);
+        table.expireSnapshots().expireOlderThan(expireTimestamp).cleanExpiredFiles(false).commit();
+        Set<String> validFiles = buildSetOfValidFiles(table);
+        try {
+            FileSystem fileSystem = hdfsEnvironment.getFileSystem(new HdfsEnvironment.HdfsContext(session), new Path(table.location()));
+            Sets.SetView<String> filesToDelete = difference(originalFiles, validFiles);
+            for (String filePath : filesToDelete) {
+                fileSystem.delete(new Path(filePath), true);
+                log.debug("%s file deleted while expiring snapshots %s", filePath, schemaTableName.getTableName());
+            }
+        }
+        catch (IOException e) {
+            throw new TrinoException(ICEBERG_FILESYSTEM_ERROR, "Failed accessing data for table: " + schemaTableName, e);
+        }
+    }
+
+    private Set<String> buildSetOfValidFiles(Table table)
```

Comment:
Out of everything, this might be the one method worth unit testing but the integration tests are pretty good if you don't want to.